### ISSUE TITLE
lets-plot in marimo: the import * problem

### ISSUE THREAD

**Issue by VivaldoMendes** [JetBrains/lets-plot#1126]
Hi,

I need to use a reactive notebook in Python, so I am playing with `marimo.py`. I also tried producing some plots using `lets-plot`.  

Unfortunately, when I tried the first line, marimo spells out:

` import * is not allowed in marimo.`

Do you have an example of how to proceed with lets-plot in order to avoid the import star problem? 

Thanks

**Comment by alshan** [comment]
To work around marimo restrictions you can:

Import the lets_plot module and then use dot-notation:
```
import lets_plot as pl

pl.LetsPlot.setup_html()

pl.ggplot(...
```

2. import symbols manually as necessary:
```
from lets_plot import ggplot, aes, geom_point, ggtitle 
```

3. Import all symbols from lets-plot using the importlib facility:
```
import lets_plot
import importlib

# Get the __all__ list from lets_plot
all_symbols = lets_plot.__all__

# Import each symbol into the current namespace
for symbol in all_symbols:
    globals()[symbol] = getattr(lets_plot, symbol)
```
This is equivalent to a "star" import.

----

However, lets-plot doesn't seem to work well in marimo notebooks out-of-the-box, according to this report: #1018.
Unfortunately, we haven't been able to take a closer look at marimo yet.

**Comment by VivaldoMendes** [comment]
Thanks for your help.

I tried as you suggested and encountered a problem that I had found in a previous attempt of my own: the code blocks run without any error being spat out, but **_no plot is visible_** (or produced?). Please see the image below:

![a](https://github.com/JetBrains/lets-plot/assets/48516539/fbfe97c7-9309-4fa6-9369-1fc583764bf0)

**Comment by alshan** [comment]
You either need to have `xix` as the last expression in the cell or call `xix.show()`.

**Comment by alshan** [comment]
Apparently marimo has its own system for handling and displaying outputs, which differы from traditional Jupyter notebooks.
AFAIU for some commonly used libraries, marimo team includes built-in support out of the box as well as creates adapters or plugins for popular visualization libraries.
Maybe a good idea would be to ask marimo to create such an adapter for lets-plot.

**Comment by VivaldoMendes** [comment]
Thank you very much for your help. You have developed a very nice package. I will pass that information to them.

**Comment by signup2k** [comment]
@VivaldoMendes 

try this, which works in my mamrio.
```
p = lp.ggplot(data, lp.aes(x='trade_date', y='price', color='symbol')) + lp.geom_line()
fig = p.to_html(iframe=True) # This is important, if you want the rendering as html
ui.Html(fig)
```

**Comment by VivaldoMendes** [comment]
@signup2k, Thanks. Unfortunately, your code does not work in my case.  I get the following error:

```
NameError: name 'ui' is not defined
```

If I remove `ui.Html(fig)` from your code, I do not get any error, but the plot is nowhere to be seen, which was what I used to get on my first attempt with lets-plot in marimo.

**Comment by signup2k** [comment]
> @signup2k, Thanks. Unfortunately, your code does not work in my case. I get the following error:
> 
> ```
> NameError: name 'ui' is not defined
> ```
> 
> If I remove `ui.Html(fig)` from your code, I do not get any error, but the plot is nowhere to be seen, which was what I used to get on my first attempt with lets-plot in marimo.

I am so sorry for this typo...It's actually `mo.Html(fig)`, where `mo` comes from `import marimo as mo`


![image](https://github.com/user-attachments/assets/84fec5d1-62bd-4e3d-90b7-b3b31a4c40a3)

**Comment by VivaldoMendes** [comment]
I'm sorry, but your code doesn't work. Can you provide a full-fledged MWE that I can use to replicate it?

**Comment by signup2k** [comment]
> I'm sorry, but your code doesn't work. Can you provide a full-fledged MWE that I can use to replicate it?

actually, it's quite simple...

```
import marimo as mo
import numpy as np
import importlib
import pandas as pd

import lets_plot as lp
lp.LetsPlot.setup_html()

# create dataframe
# create date
dates = pd.date_range(start='2024-01-01', end='2024-01-15')

# add symbols
symbols = ['MSFT', 'NVDA', 'AAPL']

# combine date and symbols
df = pd.DataFrame([(date, symbol) for date in dates for symbol in symbols], 
                  columns=['trade_date', 'symbol'])

# create random prie
np.random.seed(42)  # set random seeds
for symbol in symbols:
    start_price = np.random.uniform(100, 500)  # random initial price
    prices = np.random.normal(loc=start_price, scale=start_price*0.01, size=len(dates))
    prices = np.maximum(prices, 0)  # non-negative price
    df.loc[df['symbol'] == symbol, 'price'] = np.round(prices, 2)

# order
df = df.sort_values(['trade_date', 'symbol']).reset_index(drop=True)


# chart with lets-plot

def chart_with_lp(df):

    p = lp.ggplot(df, lp.aes(x='trade_date', y='price', color='symbol')) \
        + lp.geom_line() \
        + lp.ggtitle('Price')

    return p.to_html(iframe=True)

# show in marimo
fig = chart_with_lp(df)
mo.Html(fig)
```

**Comment by signup2k** [comment]
If the `mo.Html(fig)`  still not work. (whichi shows nothing but a blank white canvas.)

You might need to check the browser with marimo running,  whether it correctly load `lets-plot.min.js`

**Comment by VivaldoMendes** [comment]
@signup2k , thank you very much for your help. It works very well. 

Sorry for bothering you again, but how do I import the function `ggsave` to save a plot? I checked the docs, and they suggest that the use of this function would be immediate. I followed the code there, tried to save a plot, and I got this error:

```
NameError: name 'ggsave' is not defined
```
Thanks again.

**Comment by signup2k** [comment]
@VivaldoMendes 

Since, we `import lets_plot as lp` (because, marimo does not support `from lets_plot import *`, you shoud add `lp.` before `ggsave` (any call from lets_plot, should add `lp.`. 

Try `lp.ggsave()`

**Comment by VivaldoMendes** [comment]
Thanks for helping. I had tried `lp.ggsave()` before, but it did not work. Maybe I am forgetting something easy, but following the docs: `lp.ggsave(fig, 'plot.pdf', w=8, h=4, unit='in', dpi=300)`,

I got the error:

```
raise ValueError("PlotSpec, SupPlotsSpec or GGBunch expected but was: {}".format(type(plot)))
ValueError: PlotSpec, SupPlotsSpec or GGBunch expected but was: <class 'str'>
```

**Comment by alshan** [comment]
@VivaldoMendes 
In the example code above, the `fig` variable is a Python string after `p.to_html(iframe=True)`.
Whereas the 1st parameter in `ggsave()` should be `p` (again, in terms of the example above).

**Comment by VivaldoMendes** [comment]
I'm sorry, but I'm a little bit lost. The code above produces the plot; it is visible in marimo. But to save the plot as a pdf, if  `p` is defined inside the function `chart_with_lp()`, how can I call it outside that function by using that simple `lp.ggsave()`?

Thanks for the feedback.

![a](https://github.com/user-attachments/assets/7027bcfe-4096-4444-86b9-90fb6174f9a4)

**Comment by alshan** [comment]
You need to modify this code. For example, let `chart_with_lp(df)` return `p` and add  `show_lp_in_marimo(p)` function where you can do 
```
mo.Html(p.to_html(iframe=True))
```

**Comment by VivaldoMendes** [comment]
Hi @alshan and @signup2k, I now know what my problem was: I did not have the CairoSVG library on my Python installation. I had Kaleido and could export any format (pdf, SVG, png) when using other plotting packages. So, I could not figure out that I needed this particular package.

With the most recent marimo version (0.8.2), when I tried to save a plot produced by lets-plot as a PDF file, marimo spit out the following alert:

```
To export Lets-Plot figure to a PNG or PDF file please install CairoSVG libraryto your Python environment.
CairoSVG is free and distributed under the LGPL-3.0 license.
For more details visit: https://cairosvg.org/documentation/
```

Problem is solved. Thanks for helping.
